pkg/container/immset.go

```diff
@@ -68,16 +76,50 @@ func (s ImmSet[T]) Delete(xs ...T) ImmSet[T] {
 }
 
 func (s ImmSet[T]) Union(s2 ImmSet[T]) ImmSet[T] {
-	result := make([]T, len(s.xs)+len(s2.xs))
-	copy(result, s.xs)
-	copy(result[len(s.xs):], s2.xs)
-	slices.SortFunc(s.xs, s.cmp)
-	result = slices.CompactFunc(result, s.eq)
+	result := make([]T, 0, len(s.xs)+len(s2.xs))
+	i, j := 0, 0
+	for i < len(s.xs) && j < len(s2.xs) {
+		switch diff := s.cmp(s.xs[i], s2.xs[j]); {
+		case diff < 0:
+			result = append(result, s.xs[i])
+			i++
+		case diff > 0:
+			result = append(result, s2.xs[j])
+			j++
+		default:
+			result = append(result, s.xs[i])
+			i++
+			j++
+		}
+	}
+	for ; i < len(s.xs); i++ {
+		result = append(result, s.xs[i])
+	}
+	for ; j < len(s2.xs); j++ {
+		result = append(result, s2.xs[j])
+	}
 	return ImmSet[T]{result, s.cmp, s.eq}
 }
 
 func (s ImmSet[T]) Difference(s2 ImmSet[T]) ImmSet[T] {
-	return s.Delete(s2.xs...)
+	result := make([]T, 0, len(s.xs))
+	i, j := 0, 0
+	for i < len(s.xs) && j < len(s2.xs) {
+		switch diff := s.cmp(s.xs[i], s2.xs[j]); {
+		case diff < 0:
+			result = append(result, s.xs[i])
+			i++
+		case diff > 0:
+			j++
+		default:
+			i++
+			j++
+		}
+	}
+	for ; i < len(s.xs); i++ {
```

Comment:
`result = append(result, s.xs[i:]...)` should do the same.

Reply:
Of course, fixed!